runtime/src/main/java/org/corfudb/util/CFUtils.java

```diff
@@ -145,6 +150,37 @@ public static void runAfter(Duration duration, Runnable toRun) {
         return CompletableFuture.allOf(futuresArr);
     }
 
+    /**
+     * Takes a list of the completable futures and returns the CompletableFuture of a list.
+     *
+     * @param futures A list of completable futures, perhaps a result of a map function.
+     * @param <T>     A return type of the future.
+     * @return A completable future, which completes with a list of the results.
+     */
+    public static <T> CompletableFuture<List<T>> sequence(List<CompletableFuture<T>> futures) {
+        return CompletableFuture
+                .allOf(futures.toArray(new CompletableFuture<?>[futures.size()]))
+                .thenApply(x -> futures.stream()
+                        .map(CompletableFuture::join)
+                        .collect(Collectors.toList()));
+    }
+
+    public static <T> CompletableFuture<T> runFutureAfter(Supplier<CompletableFuture<T>> future,
+                                                          ScheduledExecutorService scheduledExecutorService,
```

Comment:
Please replace by non-blocking version
```
public static <T> CompletableFuture<T> runFutureAfter(
            Supplier<T> action,
            ScheduledExecutorService scheduledExecutorService,
            Duration timeout) {

        CompletableFuture<T> delayedFuture = new CompletableFuture<>();

        Runnable delayedAction = () -> {
            try {
                delayedFuture.complete(action.get());
            } catch (Exception ex) {
                delayedFuture.completeExceptionally(ex);
            }
        };
        
        scheduledExecutorService.schedule(delayedAction, timeout.toMillis(), TimeUnit.MILLISECONDS);

        return delayedFuture;
    }
```

Regarding changes:
 - I don't see reasons for using `Supplier<CompletableFuture<T>> future`, it can be simpler `Supplier<T> future`. But it's up to you.
 - passing Duration instead of long makes your code more resilient
 - also found a useful article https://www.artificialworlds.net/blog/2019/04/05/scheduling-a-task-in-java-within-a-completablefuture/

Reply:
K